File: Java/Subsets-II.java
class Solution {
    /**
     * @param S: A set of numbers.
     * @return: A list of lists. All valid subsets.
     */
    public void helper(ArrayList<ArrayList<Integer>> res, ArrayList<Integer> cur, int index, ArrayList<Integer> S) {
        res.add(new ArrayList<Integer>(cur));
        int flag = -1, prev = -1;
        for (int i = index; i < S.size(); i ++) {
            if (flag != -1 && S.get(i) == prev) continue;
            flag = 1;
            prev = S.get(i);
            cur.add(S.get(i));
            helper(res, cur, i + 1, S);
            cur.remove(cur.size() - 1);
        }
        //helper(res, cur, index + 1, S);
    }
    public ArrayList<ArrayList<Integer>> subsetsWithDup(ArrayList<Integer> S) {
        // write your code here
        ArrayList<ArrayList<Integer>> res = new ArrayList<ArrayList<Integer>>();
        if (S == null || S.size() == 0) return res;
        ArrayList<Integer> cur = new ArrayList<Integer>();
        Collections.sort(S);
        helper(res, cur, 0, S);
        return res;
    }
}

File: Python/Subsets-II.py
class Solution:
    """
    @param S: A set of numbers.
    @return: A list of lists. All valid subsets.
    """
    def helper(self, S, pos, res, cur):
        res.append(cur[:])
        pre = -1
        sign = -1
        for i in xrange(pos, len(S)):
            if sign == -1 or (sign == 1 and S[i] != pre):
                pre = S[i]
                sign = 1
                cur.append(S[i])
                self.helper(S, i + 1, res, cur)
                cur.pop()

    def subsetsWithDup(self, S):
        # write your code here
        S.sort()
        res = []
        cur = []
        self.helper(S, 0, res, cur)
        return res